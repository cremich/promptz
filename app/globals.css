pre,
code {
  word-wrap: break-word !important;
  white-space: pre-wrap !important;
  white-space: -moz-pre-wrap !important;
  white-space: -pre-wrap !important;
  white-space: -o-pre-wrap !important;
  margin: 0;
  padding: 0;
}

code {
  background-color: #f3f3f7;
}

pre > code {
  display: inline-block;
  padding: 1em;
}
