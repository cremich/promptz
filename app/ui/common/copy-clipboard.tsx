"use client";
import { Button } from "@/components/ui/button";
import { Copy } from "lucide-react";
import { toast } from "sonner";

export default function CopyClipBoardButton({
  text,
  showButtonText = false,
}: {
  text: string;
  showButtonText?: boolean;
}) {
  const copyToClipboard = async () => {
    try {
      await navigator.clipboard.writeText(text);
      toast("Copied.", {
        description: "Now, go build.",
      });
    } catch {
      toast("Failed to copy", {
        description: "Please try again",
      });
    }
  };

  return (
    <Button
      variant="outline"
      size="sm"
      className="border-gray-800 bg-transparent text-white hover:bg-violet-700"
      onClick={copyToClipboard}
    >
      <Copy className="h-4 w-4" />
      {showButtonText && <span>Copy Prompt</span>}
    </Button>
  );
}
