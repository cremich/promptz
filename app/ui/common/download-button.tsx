"use client";

import { Button } from "@/components/ui/button";
import { Download } from "lucide-react";
import { useState } from "react";

interface DownloadButtonProps {
  /**
   * The content to be downloaded
   */
  content: string;

  /**
   * The filename for the downloaded file
   */
  filename: string;

  /**
   * Optional variant for the button styling
   */
  variant?:
    | "default"
    | "destructive"
    | "outline"
    | "secondary"
    | "ghost"
    | "link";

  /**
   * Optional size for the button
   */
  size?: "default" | "sm" | "lg" | "icon";

  /**
   * Optional label for the button
   */
  label?: string;
}

/**
 * A button component that triggers a download of content as a file
 */
export function DownloadButton({
  content,
  filename,
  variant = "outline",
  size = "sm",
  label = "Download",
}: DownloadButtonProps) {
  const [isDownloading, setIsDownloading] = useState(false);

  /**
   * Handles the click event to download the content
   */
  const handleDownload = () => {
    setIsDownloading(true);

    try {
      // Create a blob with the content
      const blob = new Blob([content], { type: "text/markdown" });

      // Create a URL for the blob
      const url = URL.createObjectURL(blob);

      // Create a temporary anchor element
      const link = document.createElement("a");
      link.href = url;
      link.download = filename.endsWith(".md") ? filename : `${filename}.md`;

      // Append to the document, click it, and remove it
      document.body.appendChild(link);
      link.click();
      document.body.removeChild(link);

      // Release the object URL
      URL.revokeObjectURL(url);
    } catch (error) {
      console.error("Error downloading file:", error);
    } finally {
      setIsDownloading(false);
    }
  };

  return (
    <Button
      variant={variant}
      size={size}
      onClick={handleDownload}
      disabled={isDownloading}
      aria-label={`Download ${filename}`}
    >
      <Download className="h-4 w-4 mr-2" />
      {label}
    </Button>
  );
}
