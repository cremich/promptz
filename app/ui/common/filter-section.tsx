"use client";
import { Checkbox } from "@/components/ui/checkbox";
import {
  Collapsible,
  CollapsibleTrigger,
  CollapsibleContent,
} from "@radix-ui/react-collapsible";
import { ChevronDown } from "lucide-react";
import { useSearchParams, usePathname, useRouter } from "next/navigation";

type FilterKey = "interface" | "category" | "sdlc" | "tags";

interface FilterSectionProps {
  title: string;
  filterKey: FilterKey;
  options: string[];
  collapsible?: boolean;
}

export function FilterSection({
  title,
  filterKey,
  options,
  collapsible = false,
}: FilterSectionProps) {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const router = useRouter();

  const getSelectedValues = (filterKey: FilterKey): string[] => {
    return searchParams.getAll(`${filterKey}[]`);
  };

  const handleFilterChange = (filterKey: FilterKey, value: string) => {
    const params = new URLSearchParams(searchParams);
    const selectedValues = getSelectedValues(filterKey);

    params.delete(`${filterKey}[]`);

    if (selectedValues.includes(value)) {
      selectedValues
        .filter((v) => v !== value)
        .forEach((v) => params.append(`${filterKey}[]`, v));
    } else {
      [...selectedValues, value].forEach((v) =>
        params.append(`${filterKey}[]`, v),
      );
    }

    router.replace(`${pathname}?${params.toString()}`);
  };

  const content = (
    <div className="space-y-2">
      {options.map((option) => (
        <div key={option} className="flex items-center space-x-2">
          <Checkbox
            id={`${filterKey}-${option.toLowerCase()}`}
            checked={getSelectedValues(filterKey).includes(option)}
            onCheckedChange={() => handleFilterChange(filterKey, option)}
          />
          <label
            htmlFor={`${filterKey}-${option.toLowerCase()}`}
            className="text-sm cursor-pointer"
          >
            {option}
          </label>
        </div>
      ))}
    </div>
  );

  if (collapsible) {
    return (
      <Collapsible>
        <CollapsibleTrigger className="flex items-center justify-between w-full">
          <h3 className="font-medium">{title}</h3>
          <ChevronDown className="h-4 w-4" />
        </CollapsibleTrigger>
        <CollapsibleContent className="pt-2">{content}</CollapsibleContent>
      </Collapsible>
    );
  }

  return (
    <div>
      <h3 className="font-medium mb-3">{title}</h3>
      {content}
    </div>
  );
}
