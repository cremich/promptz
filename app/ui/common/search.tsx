"use client";
import { Input } from "@/components/ui/input";
import { Search } from "lucide-react";
import { useSearchParams, usePathname, useRouter } from "next/navigation";
import { useCallback } from "react";
import debounce from "lodash/debounce";

export default function SearchBox({ placeholder }: { placeholder: string }) {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const router = useRouter();

  // Create a debounced version of the search handler
  const debouncedSearch = useCallback(
    debounce((term: string) => {
      const params = new URLSearchParams(searchParams);

      if (term) {
        params.set("query", term);
      } else {
        params.delete("query");
      }

      router.replace(`${pathname}?${params.toString()}`);
    }, 300),
    [pathname, searchParams],
  );

  return (
    <div className="relative flex-1">
      <Search className="absolute left-3 top-1/2 h-4 w-4 -translate-y-1/2 text-muted-foreground" />
      <Input
        className="pl-10 bg-background"
        placeholder={placeholder}
        defaultValue={searchParams.get("query")?.toString()}
        onChange={(e) => debouncedSearch(e.target.value)}
      />
    </div>
  );
}
