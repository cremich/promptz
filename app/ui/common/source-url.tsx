import { Link as LinkIcon } from "lucide-react";

import { Label } from "@/components/ui/label";
import Link from "next/link";

interface SourceURLProps {
  url: string;
}

export function SourceURL({ url }: SourceURLProps) {
  return (
    <div className="flex items-center gap-2">
      <Label className="text-sm font-medium text-neutral-400">Source:</Label>
      <Link
        href={url}
        target="_blank"
        rel="noopener noreferrer"
        aria-label="View source"
        className="flex items-center gap-1.5 text-sm text-violet-400 hover:text-violet-300 transition-colors"
      >
        <LinkIcon className="h-4 w-4" />
        {url}
      </Link>
    </div>
  );
}
