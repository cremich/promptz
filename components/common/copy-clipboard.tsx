"use client";
import { Schema } from "../../amplify/data/resource";
import { ModelType } from "@/lib/forms/schema-definitions";
import { Button } from "@/components/ui/button";
import { generateClient } from "aws-amplify/api";
import { Copy } from "lucide-react";
import { toast } from "sonner";

const api = generateClient<Schema>();

export default function CopyClipBoardButton({
  id,
  text,
  type,
  showButtonText = false,
}: {
  id: string;
  text: string;
  type: ModelType;
  showButtonText?: boolean;
}) {
  const copyToClipboard = async () => {
    try {
      const clipboardPromise = navigator.clipboard.writeText(text);
      let apiPromise;
      if (type === ModelType.PROMPT) {
        apiPromise = api.mutations.copyPrompt({
          id,
        });
      } else if (type === ModelType.RULE) {
        apiPromise = api.mutations.copyProjectRule({
          id,
        });
      }
      Promise.allSettled([clipboardPromise, apiPromise]);
      toast("Copied.", {
        description: "Now, go build.",
      });
    } catch {
      toast("Failed to copy", {
        description: "Please try again",
      });
    }
  };

  return (
    <Button
      variant="outline"
      size="sm"
      className="border-gray-800 bg-transparent text-white hover:bg-violet-700"
      onClick={copyToClipboard}
    >
      <Copy className="h-4 w-4" />
      {showButtonText && <span>Copy {type}</span>}
    </Button>
  );
}
