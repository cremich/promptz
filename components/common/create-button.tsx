import { Button } from "@/components/ui/button";
import { Plus } from "lucide-react";
import Link from "next/link";

interface CreateButtonProps {
  href: string;
  name: string;
}

export default function CreateButton(props: CreateButtonProps) {
  return (
    <Button asChild className="bg-violet-500 hover:bg-violet-600">
      <Link href={props.href}>
        <Plus className="mr-2 h-4 w-4" />
        {props.name}
      </Link>
    </Button>
  );
}
