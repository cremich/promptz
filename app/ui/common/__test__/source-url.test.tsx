import { SourceURL } from "@/app/ui/common/source-url";
import { describe, expect, test } from "@jest/globals";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";

describe("SourceURL", () => {
  test("renders the prompt form component unchanged", () => {
    const { container } = render(<SourceURL url="https://promptz.dev" />);
    expect(container).toMatchSnapshot();
  });
});
