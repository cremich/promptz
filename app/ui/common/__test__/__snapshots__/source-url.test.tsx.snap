// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceURL renders the prompt form component unchanged 1`] = `
<div>
  <div
    class="flex items-center gap-2"
  >
    <label
      class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-sm font-medium text-neutral-400"
    >
      Source:
    </label>
    <a
      aria-label="View source"
      class="flex items-center gap-1.5 text-sm text-violet-400 hover:text-violet-300 transition-colors"
      href="https://promptz.dev"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        class="lucide lucide-link h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"
        />
        <path
          d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"
        />
      </svg>
      https://promptz.dev
    </a>
  </div>
</div>
`;
